### ISSUE TITLE
App not opening, when app is in background state and accepting video call from notification

### ISSUE THREAD

**Issue by nikhil-raj248** [ZEGOCLOUD/zego_uikit_prebuilt_call_flutter#80]
- [x] I have read the [Getting Started](https://www.zegocloud.com/doc/callkit/Call%20Kit/main/Quick%20start/b26d509e) and [Getting Started(with call invitation)](https://www.zegocloud.com/doc/callkit/Call%20Kit/main/Quick%20start%20(with%20call%20invitation)/eb1d3c42)
- [x] I have already searched for the same problem

## Environment

| Technology            | Version       |
|-----------------------| ------------- |
| Flutter version       |      3.24.3         |
| Plugin version        |        ^4.16.12       |
| Android version       |      14         |
| iOS version           |       17.6.1        |
| macOS version         |               |
| Xcode version         |               |
| Google Chrome version |               |

Device information: 
Android - POCO X5 (model - 22111317PI)
iOS - iPhone 13 (model - MLPK3HN/A)

## Description
I have removed all 3 permissions **(camera, microphone, systemAlertWindow)** in ZegoCallInvitationConfig when init. And handling camera and microphone permissions in  **onIncomingCallAcceptButtonPressed**  so that when call is accepted, it takes permission instead of during init of **ZegoUIKitPrebuiltCallInvitationService**, for better user experience.

**Now After this video call works fine when App is  kept in foreground or terminated state but when kept in Background state, and on accepting call from notification, App does't open by itself. Although if i explicitly open app, video call works fine but app should open by itself when call is accepted.**

Also if i take systemAlertWindow permission initially, it will work fine (for android only). But i don't want to take permissions initially and how systemAlertWindow is affecting this behavious is not clear to me.

i have tried Native methods as well in **onIncomingCallAcceptButtonPressed** to bring app in foreground state but none of them worked.

**Expected behavior:** 
when permission aren't taken initially and video call is accepted from notification when app is in background state, it should automatically come to foreground state.

**Current behavior:** 

App is not coming to foreground state, when video call is accepted from notification and app is in background state.

## Steps to reproduce
1. Remove all permission taken intially (camera, microphone, systemAlertWindow) from ZegoCallInvitationConfig of config in init of ZegoUIKitPrebuiltCallInvitationService.
2. Keep app in background state.
3. Video call notification comes and accpeted.
4. It not open app (bring to foreground state) 

![Screenshot 2024-11-24 141353](https://github.com/user-attachments/assets/ac76fe74-f76d-4b87-bad5-602d2346848b)
![Screenshot 2024-11-24 141406](https://github.com/user-attachments/assets/dbb7d398-0d24-456e-b2c0-2f8ae8d6bf03)

**Comment by phluidaungyezaw** [comment]
any solution ? @nikhil-raj248 . I also facing same issue.

**Comment by nikhil-raj248** [comment]
@yoer is it mandatory to take **systemAlertWindow** permission? and whats the fix, im still encountering the bug in both android (android 14) and ios when app kept in background state.

**Comment by yoer** [comment]
1. If you want full screen and background mode pop-up boxes, you need this permission
2. Please confirm that the correct version is used
3. If you ban these two permissions, what do you do with prebuilt call?
4. As mentioned in 1

**Comment by nikhil-raj248** [comment]
1. are you referring to incoming call notification (accept/decline) with background mode pop-up boxes? 
2. Yeah it's ^4.16.22 only
3. I want ask these permission once call is accepted. 
4. I'm checking docs that says if u need full screen then this permission is not required, in anyways full screen is not fully working - first it shows notification and then on clicking it to full screen. 

Please help @yoer 

@phluidaungyezaw can u also check the latest version.

**Comment by yoer** [comment]
1. Not only this, but also the pop-up frame for lock screen
2. If this is the case, you need to provide the log of the prebuilt call. The method is in [this document] (https://zegocloud.feishu.cn/wiki/GOEfwJX11ifNF3kTFH8c57ownCf) 2. Collect user device SDK logs chapter
3. Yes
3.1. Remove camera/microphone permissions when service.init
``` dart
ZegoUIKitPrebuiltCallInvitationService().init(
config: ZegoCallInvitationConfig(
permissions:const [
// ZegoCallInvitationPermission.camera,
// ZegoCallInvitationPermission.microphone,
ZegoCallInvitationPermission.systemAlertWindow,
],
),
...
invitationEvents: ZegoUIKitPrebuiltCallInvitationEvents(
...
)
);
```
3.2. Use permission_handler to obtain device permissions by yourself in the invitationEvents
4. SystemAlertWindow must be required for full screen permissions, and some mobile phone manufacturers have some special settings, you can view them in the notification settings of the device.

**Comment by nikhil-raj248** [comment]
1 and 4. but how do whatsapp or other calling apps do without this SystemAlertWindow window, i haven't seen these apps asking for this permission and still incoming notification and accepting it(app opening) works well. Basically i don't want this SYSTEM_ALERT_WINDOW permission to be popped for users (mostly denying this on first go).

~WhatsApp smartly avoids SYSTEM_ALERT_WINDOW by leveraging push notifications, high-priority messages, and full-screen intents. This is the best approach for VoIP apps today.


Can u please look into this. @yoer

**Comment by yoer** [comment]
> 1 and 4. but how do whatsapp or other calling apps do without this SystemAlertWindow window, i haven't seen these apps asking for this permission and still incoming notification and accepting it(app opening) works well. Basically i don't want this SYSTEM_ALERT_WINDOW permission to be popped for users (mostly denying this on first go).
> 
> ~WhatsApp smartly avoids SYSTEM_ALERT_WINDOW by leveraging push notifications, high-priority messages, and full-screen intents. This is the best approach for VoIP apps today.
> 
> Can u please look into this. [@yoer](https://github.com/yoer)

Really?  but it need takes time to check it, I don't want to use SYSTEM_ALERT_WINDOW too

**Comment by nikhil-raj248** [comment]
> > 1 and 4. but how do whatsapp or other calling apps do without this SystemAlertWindow window, i haven't seen these apps asking for this permission and still incoming notification and accepting it(app opening) works well. Basically i don't want this SYSTEM_ALERT_WINDOW permission to be popped for users (mostly denying this on first go).
> > ~WhatsApp smartly avoids SYSTEM_ALERT_WINDOW by leveraging push notifications, high-priority messages, and full-screen intents. This is the best approach for VoIP apps today.
> > Can u please look into this. [@yoer](https://github.com/yoer)
> 
> Really? but it need takes time to check it, I don't want to use SYSTEM_ALERT_WINDOW too

yeah even FCM notifications in background state, brings the app to foreground without any permission. A similar approach without SYSTEM_ALERT_WINDOW permission will be helpful for our growing user base and experience. @yoer

**Comment by nikhil-raj248** [comment]
![Image](https://github.com/user-attachments/assets/5e3a0a4d-8266-463f-9927-e76e66f0e542)
![Image](https://github.com/user-attachments/assets/e63a3473-ca19-43d6-ae73-4a9db4dbb8ff)


@yoer can u pls check this _initPermissions function, after debugging i found that for above ZegoUIKitPrebuiltCallInvitationService().init code, both the 'if' statements in _initPermissions  are getting null values hence making it true, that is why its always asking permssion. I have tried different combinations of init but everytime its asking camera and microphone permission.
